time_execution/decorator.py

```diff
@@ -1,91 +1,76 @@
-        duration = round(time.time() - self.start_time, 3) * 1000
-
-        metric = {"name": self.fqn, settings.duration_field: duration, "hostname": SHORT_HOSTNAME}
-
-        origin = getattr(settings, "origin", None)
-        if origin:
-            metric["origin"] = origin
-
-        hooks = self.params.get("extra_hooks", [])
-        disable_default_hooks = self.params.get("disable_default_hooks", False)
-
-        if not disable_default_hooks:
-            hooks = settings.hooks + hooks
-
-        # Apply the registered hooks, and collect the metadata they might
-        # return to be stored with the metrics
-        metadata = _apply_hooks(
-            hooks=hooks,
-            response=self.result,
-            exception=self.get_exception(),
-            metric=metric,
-            func=self.func,
-            func_args=self.args,
-            func_kwargs=self.kwargs,
-        )
-
-        metric.update(metadata)
-        write_metric(**metric)
-
-    def get_exception(self):
-        """Retrieve the exception"""
-        if self.exc_info is None:
-            return
-
-        exc_type, exc_value, exc_tb = self.exc_info
-        if exc_value is None:
-            exc_value = exc_type()
-        if exc_value.__traceback__ is not exc_tb:
-            return exc_value.with_traceback(exc_tb)
-        return exc_value
-
-
-class time_execution_async(AsyncDecorator, time_execution):
-    pass
+@overload
+def time_execution(__wrapped: _F) -> _F:
+    """First-order (non-parametrized) decorator with the default FQN getter and hooks by default."""
+
+
+@overload
+def time_execution(
+    *,
+    get_fqn: Callable[[Any], str] = fqn_decorators.get_fqn,
+    extra_hooks: Optional[List] = None,
+    disable_default_hooks: bool = False,
+) -> Callable[[_F], _F]:
+    """
+    Second-order (parametrized) decorator.
+
+    Args:
+        get_fqn: custom FQN getter (uses `fqn-decorators` by default)
+        extra_hooks: additional hooks (next to defined in the settings)
+        disable_default_hooks: if `True`, disable the hooks set by the settings
+    """
+
+
+def time_execution(__wrapped=None, get_fqn: Callable[[Any], str] = fqn_decorators.get_fqn, **kwargs):
+    from time_execution.timed import Timed
+
+    def wrap(__wrapped: _F) -> _F:
+        fqn = get_fqn(__wrapped)
+
+        if not iscoroutinefunction(__wrapped):
+
+            @wraps(__wrapped)
+            def wrapper(*call_args, **call_kwargs):
+                with Timed(wrapped=__wrapped, call_args=call_args, call_kwargs=call_kwargs, fqn=fqn, **kwargs) as timed:
+                    timed.result = __wrapped(*call_args, **call_kwargs)
+                    return timed.result
+
+        else:
+
+            @wraps(__wrapped)
+            async def wrapper(*call_args, **call_kwargs):
+                with Timed(wrapped=__wrapped, call_args=call_args, call_kwargs=call_kwargs, fqn=fqn, **kwargs) as timed:
+                    timed.result = await __wrapped(*call_args, **call_kwargs)
```

Comment:
Instead of setting the result in the `Timed` instance, we could yield the result to the `Timed` context manager, and return later

Reply:
I don't really get it, any `yield` here would result in `wrapper()` becoming a generator (and so breaking the wrapped function)

Reply:
Good point, my initial thought was the wrong way around. Instead, I meant to use `Generator.send`, such that:
```
with Timed(wrapped=__wrapped, call_args=call_args, call_kwargs=call_kwargs, fqn=fqn, **kwargs) as timed:
                    response = await __wrapped(*call_args, **call_kwargs)
                    timed.send(response)
                    return response
```
and timed as a function (i know it's currently a class, not sure if that could also work) would be
```
@contextmanager
def timed(...):
   start_logic()
   result = yield
   end_logic()
```

Reply:
I explored this option, and it would be even uglier, something like: `timed.gen.send(result)` (because a context manager is not a generator, and the original generator would be saved under the `.gen` attribute)

Reply:
Alright let's keep it this way then